internal/engine/buildcontroller_test.go

```diff
@@ -1047,3 +1048,303 @@ func TestLogsLongResourceName(t *testing.T) {
+
+	err := f.Stop()
+	assert.NoError(t, err)
+	f.assertAllBuildsConsumed()
+}
+
+func TestDontStartBuildIfControllerAndEngineUnsynced(t *testing.T) {
+	f := newTestFixture(t)
+	defer f.TearDown()
+
+	f.b.completeBuildsManually = true
+	f.setMaxBuildSlots(3)
+
+	manA := f.newDockerBuildManifestWithBuildPath("manA", f.JoinPath("a"))
+	manB := f.newDockerBuildManifestWithBuildPath("manB", f.JoinPath("b"))
+	manifests := []model.Manifest{manA, manB}
+
+	f.StartAndWaitForInitialBuilds(manifests)
+
+	indexA := f.editFileAndWaitForManifestBuilding("manA", "a/main.go")
+
+	// deliberately de-sync engine state and build controller
+	st := f.store.LockMutableStateForTesting()
+	st.StartedBuildCount--
+	f.store.UnlockMutableState()
+
+	// this build won't start while state and build controller are out of sync
+	f.editFileAndAssertManifestNotBuilding("manB", "b/main.go")
+
+	// resync the two counts, manB build will start as expected
+	indexB := f.runAndWaitForManifestBuilding("manB", func() {
+		st := f.store.LockMutableStateForTesting()
+		st.StartedBuildCount++
+		f.store.UnlockMutableState()
+	})
+
+	// complete all builds (can't guarantee order)
+	f.completeAndCheckBuildsForManifests([]int{indexA, indexB}, manifests)
+
+	err := f.Stop()
+	assert.NoError(t, err)
+	f.assertAllBuildsConsumed()
+}
+
+func TestErrorHandlingWithMultipleBuilds(t *testing.T) {
+	f := newTestFixture(t)
+	defer f.TearDown()
+	f.b.completeBuildsManually = true
+	f.setMaxBuildSlots(2)
+
+	errA := fmt.Errorf("errA")
+	errB := fmt.Errorf("errB")
+
+	manA := f.newDockerBuildManifestWithBuildPath("manA", f.JoinPath("a"))
+	manB := f.newDockerBuildManifestWithBuildPath("manB", f.JoinPath("b"))
+	manC := f.newDockerBuildManifestWithBuildPath("manC", f.JoinPath("c"))
+	f.StartAndWaitForInitialBuilds([]model.Manifest{manA, manB, manC})
+
+	// start builds for all manifests (we only have 2 build slots)
+	f.SetNextBuildFailure(errA)
+	indexA := f.editFileAndWaitForManifestBuilding("manA", "a/main.go")
+	f.SetNextBuildFailure(errB)
+	indexB := f.editFileAndWaitForManifestBuilding("manB", "b/main.go")
+	f.editFileAndAssertManifestNotBuilding("manC", "c/main.go")
+
+	// Complete one build -- 'manC' should start building, even though the manA build ended with an error
+	indexC := f.runAndWaitForManifestBuilding("manC", func() {
+		f.b.completeBuild(indexA)
+		call := f.nextCall("expect manA build complete")
+		f.assertCallIsForManifestAndFiles(call, manA, "a/main.go")
+		f.WaitUntilManifestState("last manA build reflects expected error", "manA", func(ms store.ManifestState) bool {
+			return ms.LastBuild().Error == errA
+		})
+	})
+
+	// complete the rest
+	f.completeAndCheckBuildsForManifests([]int{indexB, indexC}, []model.Manifest{manB, manC})
+	f.WaitUntilManifestState("last manB build reflects expected error", "manB", func(ms store.ManifestState) bool {
+		return ms.LastBuild().Error == errB
+	})
+	time.Sleep(10 * time.Millisecond)
```

Comment:
it takes a few ms for the BuildComplete action to bubble through, so we could potentially get a false positive if we just checked for "manifest C last build has no error" before a BuildComplete action had made it through the pipe. (Granted this is LESS likely because we know the manifest B BuildComplete has already made it through, but seemed prudent.)

will add a comment, unless you think this is unnecessary

Reply:
hmmm...but that means if the behavior is totally broken, then this test will pass? right?

i'm not opposed to adding Sleeps(). But in this case, the sleep is a code smell that the test has deeper problems.